src/ui/public/test_harness/test_sharding/root_describe_call_filter.js

```diff
@@ -0,0 +1,90 @@
+/**
+ *  Intercept all calls to mocha.describe() and determine
+ *  which calls make it through using a filter function.
+ *
+ *  The filter function is also only called for ROOT describe()
+ *  calls, and all describe calls inside any allowed call are
+ *  automatically passed through.
+ *
+ *  ## example
+ *
+ *  assume tests that look like this:
+ *
+ *  ```js
+ *  describe('section 1', () => {
+ *    describe('item 1', () => {
+ *
+ *    })
+ *  })
+ *  ```
+ *
+ *  If the filter function returned true for "section 1" then "item 1"
+ *  would automatically be defined. If it returned false for "section 1"
+ *  then "section 1" would be ignored and "item 1" would never be defined
+ *
+ *  @param {function} test - a function that takes the first argument
+ *                             passed to describe, the sections name
+ *  @return {undefined}
+ */
+export function setupRootDescribeCallFilter(test) {
+  const originalDescribe = window.describe;
+
+  if (!originalDescribe) {
+    throw new TypeError('window.describe must be defined by mocha before test sharding can be setup');
+  }
+
+  /**
+   *  When describe is called it is likely to make additional, nested,
+   *  calls to describe. We track how deeply nested we are at any time
+   *  with a depth counter, `describeCallDepth`.
+   *
+   *  Before delegating a describe call to mocha we increment
+   *  that counter, and once mocha is done we decrement it.
+   *
+   *  This way, we can check if `describeCallDepth > 0` at any time
+   *  to know if we are already within a describe call or not
+   *
+   *  ```js
+   *  // +1
+   *  describe('section 1', () => {
+   *    // describeCallDepth = 1
+   *    // +1
+   *    describe('item 1', () => {
+   *      // describeCallDepth = 2
+   *    })
+   *    // -1
+   *  })
+   *  // -1
+   *  // describeCallDepth = 0
+   *  ```
+   *
+   *  @type {Number}
+   */
+  let describeCallDepth = 0;
+  const ignoredCalls = [];
+
+  // ensure that window.describe isn't messed with by other code,
+  // if it becomes necessary in the future we can make this fancier
+  Object.defineProperty(window, 'describe', {
```

Comment:
There is a chance that something will try to override window.describe, but since the describe property is now `configurable: false` and `writable: false` (the default) it will cause an error to be thrown if something tries to change it.